File: pa1/rosetta.js
var readline = require('readline');

var rl = readline.createInterface({
	input: process.stdin,
	output: process.stdout,
	terminal: false
});

var graph = {};
var indegree = {};
var start = [];
var lines = [];
var cycle = false;

// Read input, called on '\n' characters
rl.on('line', function(line) {
	lines.push(line)
});

// Main body
rl.on('close', function() {

	var dst = "";
	var src = "";

	// Construct graph from buffered input
	for (i = 0; i < lines.length; i += 2) {
		dst = lines[i];
		src = lines[i+1];

		// Add nodes to graph if newly encountered
		if (!(src in graph)) {
			graph[src] = [];
			indegree[src] = 0;
		}

		if (!(dst in graph)) {
			graph[dst] = [];
			indegree[dst] = 0;
		}

		// Add edge to graph
		graph[src].push(dst);

		// Increment indegree of dst
		indegree[dst] = indegree[dst] + 1;
	}

	// Construct start set
	for (key in indegree) {
		if (indegree[key] == 0) {
			start.push(key);
		}
	}

	// No nodes w/ indegree zero -> cycle
	if (start.length == 0) {
		cycle = true;
	}

	// Top sort
	var sorted = []

	while (start.length > 0) {
		// Get minimum node from start set
		src = start[0];
		start.forEach(function(node) {
			if (src.localeCompare(node) > 0) {
				src = node;
			}
		});

		// Remove node from start set, append to sorted list
		start.splice(start.indexOf(src), 1);
		sorted.push(src);

		// Decrement indegrees of adjacent nodes
		adj = graph[src];

		if (adj.length > 0) {
			adj.forEach(function(dst) {
				indegree[dst] = indegree[dst] - 1;

				// Check for new start nodes
				if (indegree[dst] == 0) {
					start.push(dst);
				}
			});
		}

	}

	// Non-empty graph -> cycle
	for (key in indegree) {
		if (indegree[key] != 0) {
			cycle = true;
		}
	}

	// Output
	if (cycle) {
		process.stdout.write("cycle\n");
	} else {
		sorted.forEach(function(node) {
			process.stdout.write(node + "\n");
		});
	}
});

File: pa1/rosetta.rb
require 'set'

graph = Hash.new()
indegree = Hash.new()
start = Set.new()
cycle = false

while true
	# Get src/dst from stdin
	dst = gets
	if dst == nil
		break
	end
	src = gets

	dst.strip!
	src.strip!

	# Add nodes to graph/indegrees/start set
	if !graph.has_key?(dst)
		graph[dst] = []
		indegree[dst] = 0
		start << dst
	end #endif


	if !graph.has_key?(src)
		graph[src] = []
		indegree[src] = 0
		start << src
	end #endif

	# Add edge to graph, increment indegree of dst
	graph[src] << dst
	indegree[dst] += 1
end

# graph.each{|key,value| puts "#{key}: #{value}"}

# Get nodes with indegree zero
start.each do |node|
	if (indegree[node] != 0)
		start.delete(node)
	end
end

if start.empty?()
	cycle = true
end

sorted = []

while !start.empty?()
	# Get min
	src = start.to_a().min()

	# Remove min, add to sorted
	start.delete(src)
	sorted << src

	graph[src].each do |dst|
		indegree[dst] -= 1
		if indegree[dst] == 0
			start << dst
		end
	end
end

indegree.each_value do |value|
	if (value != 0)
		cycle = true
	end
end

if cycle
	puts "cycle"
else
	sorted.each {|n| puts "#{n}\n"}
end